File: app/src/main/java/eventcoordinator2017/myevent/ui/main/MainActivity.java
package eventcoordinator2017.myevent.ui.main;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.gms.common.api.GoogleApiClient;
import com.hannesdorfmann.mosby.mvp.MvpActivity;

import java.util.ArrayList;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import eventcoordinator2017.myevent.R;
import eventcoordinator2017.myevent.app.App;
import eventcoordinator2017.myevent.app.Constants;
import eventcoordinator2017.myevent.databinding.ActivityMainBinding;
import eventcoordinator2017.myevent.model.data.Event;
import eventcoordinator2017.myevent.model.data.User;
import eventcoordinator2017.myevent.model.response.ResultResponse;
import eventcoordinator2017.myevent.ui.events.EventsActivity;
import eventcoordinator2017.myevent.ui.login.LoginActivity;
import eventcoordinator2017.myevent.ui.notifs.NotificationsActivity;
import eventcoordinator2017.myevent.ui.profile.ProfileActivity;
import eventcoordinator2017.myevent.utils.SharedPreferencesUtil;
import io.realm.Realm;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class MainActivity extends MvpActivity<MainView, MainPresenter> implements MainView, NavigationView.OnNavigationItemSelectedListener {
    private static final String TAG = MainActivity.class.getSimpleName();
    private ActivityMainBinding binding;
    /**
     * ATTENTION: This was auto-generated to implement the App Indexing API.
     * See https://g.co/AppIndexing/AndroidStudio for more information.
     */
    private GoogleApiClient client;
    private ProgressDialog progressDialog;
    private List<String> strings = new ArrayList<>();
    private SearchView searchView;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.setView(getMvpView());
        presenter.onStart();
        setSupportActionBar(binding.toolbar);


        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, binding.toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.setDrawerListener(toggle);
        toggle.syncState();

        binding.navigationView.setNavigationItemSelectedListener(this);


        //display data
        binding.navigationView.getHeaderView(0).findViewById(R.id.viewProfile).setOnClickListener(view -> startActivity(new Intent(MainActivity.this, ProfileActivity.class)));

        binding.navigationView.getMenu().getItem(0).setChecked(true);

        strings.add("Today");
        strings.add("Upcoming");
        binding.viewpager.setAdapter(new MainPageFragmentAdapter(getSupportFragmentManager(), this, strings));
        binding.viewpager.setOffscreenPageLimit(strings.size());
        binding.slidingTabs.setupWithViewPager(binding.viewpager);
        binding.swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                presenter.getEvents();
            }
        });


        sendTokenToServer();
    }

    @Override
    protected void onResume() {
        super.onResume();
        final Realm realm = Realm.getDefaultInstance();
        User user = realm.where(User.class).findFirst();
        if (user != null)
            displayUserData(user);
        realm.close();
    }

    private void sendTokenToServer() {
        User user = App.getUser();
        final SharedPreferencesUtil sharedPreferencesUtil = new SharedPreferencesUtil(this, Constants.FIREBASE);
        Boolean sent = sharedPreferencesUtil.getBooleanValue(Constants.FIREBASE + "_sent", false);
        String token = sharedPreferencesUtil.getStringValue(Constants.FIREBASE + "_token", "");
        if (!sent) {
            if (!token.equals("")) {
                App.getInstance().getApiInterface().saveUserToken(user.getUserId() + "", token).enqueue(new Callback<ResultResponse>() {
                    @Override
                    public void onResponse(Call<ResultResponse> call, Response<ResultResponse> response) {
                        if (response.isSuccessful()) {
                            if (response.body().equals(Constants.SUCCESS)) {
                                sharedPreferencesUtil.putBooleanValue(Constants.FIREBASE + "_sent", true);
                            } else {
                                Log.e(TAG, "Token Not Updated");
                            }
                        } else {
                            showAlert("Website is sleeping");
                        }
                    }

                    @Override
                    public void onFailure(Call<ResultResponse> call, Throwable t) {
                        Log.d(TAG, t.toString());
                    }
                });
            } else {
                Log.e(TAG, "No Token Yet");
            }
        } else {
            Log.e(TAG, "Token is already sent");
        }
    }


    @Override
    public void startLoading() {
        binding.swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void stopLoading() {
        binding.swipeRefreshLayout.setRefreshing(false);
    }


    @NonNull
    @Override
    public MainPresenter createPresenter() {
        return new MainPresenter();
    }

    @Override
    public void displayUserData(User user) {
        // TextView email = (TextView) binding.navigationView.getHeaderView(0).findViewById(R.id.email);
        TextView name = binding.navigationView.getHeaderView(0).findViewById(R.id.name);
        CircleImageView circleImageView = binding.navigationView.getHeaderView(0).findViewById(R.id.userImage);
        // email.setText(user.getEmail());
        name.setText(user.getFullName());
        Glide.with(this)
                .load(Constants.URL_IMAGE + user.getImage())
                .error(R.drawable.ic_gallery)
                .into(circleImageView);
    }

    @Override
    public void showAlert(String s) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void setEvents(List<Event> eventToday, List<Event> eventUpcoming) {
        binding.viewpager.setAdapter(new MainPageFragmentAdapter(getSupportFragmentManager(), this, strings));
    }

    @Override
    public void refreshList() {

    }


    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            if (!searchView.isIconified()) {
                searchView.setIconified(true);
                return;
            }
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        /*getMenuInflater().inflate(R.menu.menu_search, menu);

        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setQueryHint("Budget, Name");

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
               // presenter.setApplyFilter(filterType, filterSort, query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                //presenter.setApplyFilter(filterType, filterSort, query);
                return false;
            }
        });*/

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public boolean onNavigationItemSelected(MenuItem item) {
        // Handle navigation view item clicks here.
        int id = item.getItemId();

        if (id == R.id.home) {

        } else if (id == R.id.events) {
            startActivity(new Intent(this, EventsActivity.class));
            binding.navigationView.getMenu().getItem(0).setChecked(true);
        } else if (id == R.id.notification) {
            startActivity(new Intent(this, NotificationsActivity.class));
            binding.navigationView.getMenu().getItem(0).setChecked(true);
        } else if (id == R.id.logout) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Log Out");
            builder.setMessage("Are you sure?");
            builder.setPositiveButton("YES", (dialog, which) -> {
                // Do nothing but close the dialog
                final Realm realm = Realm.getDefaultInstance();
                realm.executeTransactionAsync(realm1 -> realm1.deleteAll(), () -> {
                    startActivity(new Intent(MainActivity.this, LoginActivity.class).addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP));
                    MainActivity.this.finish();
                }, error -> {
                    realm.close();
                    Log.e(TAG, "onError: Error Logging out (deleting all data)", error);
                });
                finish();
            });
            builder.setNegativeButton("NO", new DialogInterface.OnClickListener() {

                @Override
                public void onClick(DialogInterface dialog, int which) {

                    // Do nothing
                    dialog.dismiss();
                }
            });
            AlertDialog alert = builder.create();
            alert.show();

        }
        DrawerLayout drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.onStop();
    }


}

File: app/src/main/java/eventcoordinator2017/myevent/ui/main/MainActivity.kt
package eventcoordinator2017.myevent.ui.main

import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.Intent
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.SearchView
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast

import com.bumptech.glide.Glide
import com.google.android.gms.common.api.GoogleApiClient
import com.hannesdorfmann.mosby.mvp.MvpActivity

import java.util.ArrayList

import de.hdodenhof.circleimageview.CircleImageView
import eventcoordinator2017.myevent.R
import eventcoordinator2017.myevent.app.App
import eventcoordinator2017.myevent.app.Constants
import eventcoordinator2017.myevent.databinding.ActivityMainBinding
import eventcoordinator2017.myevent.model.data.Event
import eventcoordinator2017.myevent.model.data.User
import eventcoordinator2017.myevent.model.response.ResultResponse
import eventcoordinator2017.myevent.ui.events.EventsActivity
import eventcoordinator2017.myevent.ui.login.LoginActivity
import eventcoordinator2017.myevent.ui.notifs.NotificationsActivity
import eventcoordinator2017.myevent.ui.profile.ProfileActivity
import eventcoordinator2017.myevent.utils.SharedPreferencesUtil
import io.realm.Realm
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


class MainActivity : MvpActivity<MainView, MainPresenter>(), MainView, NavigationView.OnNavigationItemSelectedListener {
    private lateinit var binding: ActivityMainBinding
    private lateinit var client: GoogleApiClient
    private lateinit var progressDialog: ProgressDialog
    private var strings = ArrayList<String>()
    private lateinit var searchView: SearchView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        binding.view = mvpView
        presenter.onStart()
        setSupportActionBar(binding.toolbar)


        val drawer = binding.drawerLayout
        val toggle = ActionBarDrawerToggle(
                this, drawer, binding.toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        drawer.setDrawerListener(toggle)
        toggle.syncState()

        binding.navigationView.setNavigationItemSelectedListener(this)


        //display data
        binding.navigationView.getHeaderView(0).findViewById<View>(R.id.viewProfile).setOnClickListener {
            startActivity(Intent(this@MainActivity, ProfileActivity::class.java))
        }

        binding.navigationView.menu.getItem(0).isChecked = true

        strings.add("Today")
        strings.add("Upcoming")
        binding.viewpager.apply {
            adapter = MainPageFragmentAdapter(supportFragmentManager, this.context, strings)
            offscreenPageLimit = strings.size
        }
        binding.slidingTabs.setupWithViewPager(binding.viewpager)
        binding.swipeRefreshLayout.setOnRefreshListener { presenter.getEvents() }


        //sendTokenToServer();
    }

    override fun onResume() {
        super.onResume()
        val realm = Realm.getDefaultInstance()
        val user = realm.where(User::class.java).findFirst()
        if (user != null)
            displayUserData(user)
        realm.close()
    }

    /*private fun sendTokenToServer() {
        val user = App.getUser()
        val sharedPreferencesUtil = SharedPreferencesUtil(this, Constants.FIREBASE)
        val sent = sharedPreferencesUtil.getBooleanValue(Constants.FIREBASE + "_sent", false)
        val token = sharedPreferencesUtil.getStringValue(Constants.FIREBASE + "_token", "")
        if (!sent) {
            if (token != "") {
                App.getInstance().apiInterface.saveUserToken(user.userId.toString() + "", token).enqueue(object : Callback<ResultResponse> {
                    override fun onResponse(call: Call<ResultResponse>, response: Response<ResultResponse>) {
                        if (response.isSuccessful) {
                            if (response.body() == Constants.SUCCESS) {
                                sharedPreferencesUtil.putBooleanValue(Constants.FIREBASE + "_sent", true)
                            } else {
                                Log.e(TAG, "Token Not Updated")
                            }
                        } else {
                            showAlert("Website is sleeping")
                        }
                    }

                    override fun onFailure(call: Call<ResultResponse>, t: Throwable) {
                        Log.d(TAG, t.toString())
                    }
                })
            } else {
                Log.e(TAG, "No Token Yet")
            }
        } else {
            Log.e(TAG, "Token is already sent")
        }
    }*/


    override fun startLoading() {
        binding.swipeRefreshLayout.isRefreshing = true
    }

    override fun stopLoading() {
        binding.swipeRefreshLayout.isRefreshing = false
    }


    override fun createPresenter(): MainPresenter {
        return MainPresenter()
    }

    override fun displayUserData(user: User) {
        // TextView email = (TextView) binding.navigationView.getHeaderView(0).findViewById(R.id.email);
        val name = binding.navigationView.getHeaderView(0).findViewById<TextView>(R.id.name)
        val circleImageView = binding.navigationView.getHeaderView(0).findViewById<CircleImageView>(R.id.userImage)
        // email.setText(user.getEmail());
        name.text = user.fullName
        Glide.with(this)
                .load(Constants.URL_IMAGE + user.image)
                .error(R.drawable.ic_gallery)
                .into(circleImageView)
    }

    override fun showAlert(s: String) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show()
    }

    override fun setEvents(eventToday: List<Event>, eventUpcoming: List<Event>) {
        binding.viewpager.adapter = MainPageFragmentAdapter(supportFragmentManager, this, strings)
    }

    override fun refreshList() {

    }


    override fun onBackPressed() {
        val drawer = findViewById<DrawerLayout>(R.id.drawer_layout)
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START)
        } else {
            if (!searchView.isIconified) {
                searchView.isIconified = true
                return
            }
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        /*getMenuInflater().inflate(R.menu.menu_search, menu);

        SearchManager searchManager = (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setQueryHint("Budget, Name");

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
               // presenter.setApplyFilter(filterType, filterSort, query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {
                //presenter.setApplyFilter(filterType, filterSort, query);
                return false;
            }
        });*/

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = item.itemId

        when (id) {
            R.id.home -> {

            }
            R.id.events -> {
                startActivity(Intent(this, EventsActivity::class.java))
                binding.navigationView.menu.getItem(0).isChecked = true
            }
            R.id.notification -> {
                startActivity(Intent(this, NotificationsActivity::class.java))
                binding.navigationView.menu.getItem(0).isChecked = true
            }
            R.id.logout -> {
                val builder = AlertDialog.Builder(this)
                builder.setTitle("Log Out")
                builder.setMessage("Are you sure?")
                builder.setPositiveButton("YES") { dialog, which ->
                    // Do nothing but close the dialog
                    val realm = Realm.getDefaultInstance()
                    realm.executeTransactionAsync({ realm1 -> realm1.deleteAll() }, {
                        startActivity(Intent(this@MainActivity, LoginActivity::class.java).addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP))
                        this@MainActivity.finish()
                    }, { error ->
                        realm.close()
                        Log.e(TAG, "onError: Error Logging out (deleting all data)", error)
                    })
                    finish()
                }
                builder.setNegativeButton("NO") { dialog, which ->
                    // Do nothing
                    dialog.dismiss()
                }
                val alert = builder.create()
                alert.show()

            }
        }
        val drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        drawer.closeDrawer(GravityCompat.START)
        return true
    }

    public override fun onDestroy() {
        super.onDestroy()
        presenter.onStop()
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }


}
